tests/CallbacksTest.test.nut

```diff
@@ -0,0 +1,173 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO
+// EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+// OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+@include "./ConnectionManager.lib.nut"
+
+class CallbacksTest extends ImpTestCase {
+
+  cm = null;
+
+  function setUp() {
+    this.info("running setUp");
+    cm = ConnectionManager();
+  }
+
+  /*
+  * sets onDisconnect callback and disconnects device using CM
+  *
+  */
+  function testOnDisconnectAsync() {
+    return Promise(function(resolve, reject) {
+      cm.onDisconnect(function(expected) {
+        cm.connect();
+      }.bindenv(this));
+
+      cm.onConnect(function() {
+        resolve();
+      }.bindenv(this));
+
+      cm.disconnect();
+      assertTrue(!server.isconnected(), "should NOT be connected!");
+    }.bindenv(this))
+    .then(function(val) {
+        assertTrue(server.isconnected(), "should be connected again!");
+        _resetCM.call(this);
+    }.bindenv(this))
+    .fail(function(valueOrReason) {
+        _resetCM.call(this);
+        throw valueOrReason;
+      }.bindenv(this));
+  }
+
+ /*
+  * sets onConnect callback and disconnects device using CM
+  *
+  */
+  function testOnConnectAsync() {
+    local flag = 0;
+    return Promise(function(resolve, reject) {
+      cm.onDisconnect(function(expected) {
+        assertTrue(!flag, "should be 0 now");
+        cm.connect();
+      }.bindenv(this));
+
+      cm.onConnect(function() {
+        flag++;
+        resolve();
+      }.bindenv(this));
+
+      assertTrue(!flag, "should 0 before re-connect");
+      cm.disconnect();
+      assertTrue(!server.isconnected(), "should NOT be connected!");
+    }.bindenv(this))
+    .then(function(val) {
+      //not assertTrue(flag, ...) to ensure flag was increased exactly one time
+      assertTrue(flag==1, "connected again, should be 1 now");
+      _resetCM.call(this);
+    }.bindenv(this))
+    .fail(function(valueOrReason) {
+      _resetCM.call(this);
+      throw valueOrReason;
+    }.bindenv(this));
+  }
+
+ /*
+  * sets onNextConnect callback and disconnects device using CM
+  *
+  */
+  function testOnNextConnectAsync() {
+    local flag = 0;
+    return Promise(function(resolve, reject) {
+      cm.onDisconnect(function(expected) {
+        assertTrue(!flag, "should be 0 now");
+        cm.onNextConnect(function() {
```

Comment:
Do we need to set null for the  "onNextConnect" in the _resetCM method too ?

Reply:
cm._queue = [];  - or any other way to remove queue. The same queue uses for the connectFor

Reply:
There's no null check in onNextConnect, so I'm not sure if  it's a  good  idea to pass null there.
_queue is a private field, so changing it manually will affect internal state of CM in an unpredictable way that may change during CM development since it's not a part of public API.

Reply:
Then it's better to move such test into the separate file, to avoid unexpected behavior